src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -1149,23 +1149,47 @@ public override BoundNode VisitDynamicInvocation(BoundDynamicInvocation node)
             return null;
         }
 
-        protected BoundNode VisitInterpolatedStringBase(BoundInterpolatedStringBase node)
+        protected BoundNode VisitInterpolatedStringBase(BoundInterpolatedStringBase node, InterpolatedStringHandlerData? nullableData)
         {
-            foreach (var expr in node.Parts)
+            // If there can be any branching, then we need to treat the expressions
+            // as optionally evaluated. Otherwise, we treat them as always evaluated
+            switch (nullableData)
             {
-                VisitRvalue(expr);
+                case null:
+                    visitParts();
+                    break;
+                case { HasTrailingHandlerValidityParameter: false, UsesBoolReturns: false, Construction: var construction }:
+                    VisitRvalue(construction);
+                    visitParts();
+                    break;
+                case { Construction: var construction }:
+                    VisitRvalue(construction);
+                    var beforePartsState = State.Clone();
+                    visitParts();
+                    Join(ref State, ref beforePartsState);
+                    break;
             }
```

Comment:
nit: consider adding a default case with a `throw`

Reply:
Not possible to hit. The 3 cases given are fully exhaustive.